stage0/gc_test.go

```diff
@@ -15,48 +15,130 @@
+248 242 0:32 / /prefix/stage1/rootfs/opt/stage2/busybox/rootfs/rootfs/prefix/stage1/rootfs/opt/stage2/busybox/rootfs/sys/fs/cgroup/memory rw,nosuid,nodev,noexec,relatime -
+249 242 0:33 / /prefix/stage1/rootfs/opt/stage2/busybox/rootfs/rootfs/prefix/stage1/rootfs/opt/stage2/busybox/rootfs/sys/fs/cgroup/devices rw,nosuid,nodev,noexec,relatime -
+250 242 0:34 / /prefix/stage1/rootfs/opt/stage2/busybox/rootfs/rootfs/prefix/stage1/rootfs/opt/stage2/busybox/rootfs/sys/fs/cgroup/blkio rw,nosuid,nodev,noexec,relatime -
+251 242 0:29 / /prefix/stage1/rootfs/opt/stage2/busybox/rootfs/rootfs/prefix/stage1/rootfs/opt/stage2/busybox/rootfs/sys/fs/cgroup/cros rw,relatime -
+252 234 0:47 / /prefix/stage1/rootfs/opt/stage2/busybox/rootfs/rootfs/prefix/stage1/rootfs/opt/stage2/busybox/rootfs/tmp rw,relatime -
+253 234 0:19 /nixos/etc/resolv.conf /prefix/stage1/rootfs/opt/stage2/busybox/rootfs/rootfs/prefix/stage1/rootfs/opt/stage2/busybox/rootfs/etc/resolv.conf ro,relatime -
+254 193 0:19 /home /prefix/stage1/rootfs/opt/stage2/busybox/rootfs/rootfs/home rw,relatime -
+255 193 179:1 / /prefix/stage1/rootfs/opt/stage2/busybox/rootfs/rootfs/mnt/unknown ro,relatime -
+`
 
 func TestMountOrdering(t *testing.T) {
 	tests := []struct {
 		prefix     string
 		ids        []int
 		shouldPass bool
+		mi         string
 	}{
 		{
-			prefix:     "/my/mount/prefix",
-			ids:        []int{2, 9, 8, 7, 6},
+			prefix: "/prefix",
+			ids: []int{
+				146, 145, 221, 183, 219, 218, 182, 217, 210, 209, 180, 144, 121, 116,
+			},
 			shouldPass: true,
+			mi:         mountinfo,
 		},
 	}
 
 	for i, tt := range tests {
-		mi := strings.NewReader(mountinfo)
+		mi := strings.NewReader(tt.mi)
 		mnts, err := getMountsForPrefix(tt.prefix, mi)
 		if err != nil {
 			t.Errorf("problems finding mount points: %v", err)
 		}
 
 		if len(mnts) != len(tt.ids) {
-			t.Errorf("test  %d: didn't find the expected number of mounts. found %d but wanted %d.", i, len(mnts), len(tt.ids))
-			return
+			t.Logf("test  %d: didn't find the expected number of mounts. found %d but wanted %d.", i, len(mnts), len(tt.ids))
+		} else {
+			mountIds := make([]int, len(tt.ids))
+			match := true
+			for j, mnt := range mnts {
+				mountIds[j] = mnt.id
+				match = match && (mnt.id == tt.ids[j])
+			}
+			if !match {
+				t.Logf("test #%d: unexpected mount ordering; expected\n%v, got\n%v", i, tt.ids, mountIds)
+			}
 		}
 
-		for j, mnt := range mnts {
-			if mnt.id != tt.ids[j] {
-				t.Errorf("test #%d: problem with mount ordering; mount at index %d is %d not %d",
-					i, j, mnt.id, tt.ids[j])
-				return
+		for _, mntCur := range mnts {
+			t.Logf("Unmounting %d: %q", mntCur.id, mntCur.mountPoint)
+		}
+
+		if err := mnts.SafeOrder(); err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+// SafeOrder checks for transitivity and (un)mount order sanity
+func (m mounts) SafeOrder() error {
```

Comment:
I actually agree that this should be a function rather than a method. We should have a *method or function* preference discussed and documented somewhere.